### PR TITLE
refactor: remove charm set revision

### PR DESCRIPTION
**Pull request by SimonRichardson**
The revision is set on a local charm to seed the charm archive, which is then read from the archive to populate the database. At no point in the juju code base is the revision then read from the archive instead of the database.

It's possible for a charmer to read the revision from disk, except it's not consistent from local charms to charmhub charms. If they did want to get the revision, it would be better to set the revision in a hook envvar and pass that through. It would be more "correct".

This was only possible thanks to @jack-w-shaw refactoring of the package and upload code in https://github.com/juju/juju/pull/18516

----

This now fixes the issue of charms objects endpoints adding a charm with a hash, only to be changed by the server and unable to locate the charm with the same sha without an additional call to locate the charm. Thus making the PUT and GET requests isomorphic.

----

This pull request includes several changes to the charm handling functionality, specifically focusing on the removal of the `SetRevision` method and related code, as well as refactoring to improve code clarity and functionality.

### Removal of `SetRevision` method and related code:

* [`apiserver/charms.go`](diffhunk://#diff-8fd2694ecf2a17b1bb232e39c62e09ce47898c13e6ada785db4bf1a58ad42448L1-L51): Removed the `repackageCharmWithRevision` function and its associated imports, which included logic for repackaging charms with a specific revision.
* [`internal/charm/charmarchive_test.go`](diffhunk://#diff-9e6ac1e14acee56a5bde1d3a00cd1cb8fa36677912637d47e6ad69a1c7d4a612L353-L369): Removed the `TestCharmArchiveSetRevision` test, which tested the functionality of setting a charm revision.
* [`internal/charm/charmbase.go`](diffhunk://#diff-8b8f9952f9cbb087890b42c7b0bc7a447440f49bb7d7589727c025adf5561d21L71-R83): Removed the `SetRevision` method and replaced it with a private `setRevision` method to encapsulate revision setting within the class.
* [`internal/charm/charmdir.go`](diffhunk://#diff-8987e5b00ffbacac86d1e16ec7c5e8231f7774461763baf6b5d78da62523bc0bL173-R173): Updated the `SetDiskRevision` method to use the private `setRevision` method instead of the removed `SetRevision` method.
* [`internal/charm/charmdir_test.go`](diffhunk://#diff-086f2d3f1a89d502c66146b8bc49b155ad5645f705c4ceea7d6176b7fb87aaf7L347-R347): Modified the `TestDirSetRevision` test to use `SetDiskRevision` instead of the removed `SetRevision` method.

### Refactoring of charm processing logic:

* [`apiserver/objects.go`](diffhunk://#diff-fab5b76e7c234d9c929014d46ef0a5db9c8b6e9fd63bdc3ba9c2b903471bc77eL224-L230): Removed the call to `repackageCharmWithRevision` in the `processPut` function, simplifying the charm processing logic.

### Local Controller

```
$ juju download juju-controller --channel 4.0
Fetching charm "juju-controller" revision 117 using "4.0/stable" channel and base "amd64/ubuntu/24.04"
Install the "juju-controller" charm with:
    juju deploy ./juju-controller_r117.charm

$ juju bootstrap --controller-charm-path ./juju-controller_r117.charm lxd local

$ juju status -m controller
Model       Controller  Cloud/Region         Version      Timestamp
controller  local       localhost/localhost  4.0-beta5.1  21:01:06Z

App         Version  Status  Scale  Charm            Channel  Rev  Exposed  Message
controller           active      1  juju-controller             0  no       

Unit           Workload  Agent  Machine  Public address  Ports  Message
controller/0*  active    idle   0        10.219.211.126         

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.126  juju-d7172e-0  ubuntu@24.04      Running

$ juju debug-log -m controller
(verify no unexpected errors)
```

### Deploy local charms

```
$ juju bootstrap lxd lxd
$ juju add-model m
$ juju download ubuntu
$ juju download ubuntu
Fetching charm "ubuntu" revision 25 using "stable" channel and base "amd64/ubuntu/24.04"
Install the "ubuntu" charm with:
    juju deploy ./ubuntu_r25.charm

$ juju deploy ./ubuntu_r25.charm
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m      local       localhost/localhost  4.0-beta5.1  21:03:47Z

App     Version  Status  Scale  Charm   Channel  Rev  Exposed  Message
ubuntu  24.04    active      1  ubuntu             0  no       

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   0        10.219.211.94          

Machine  State    Address        Inst id        Base          AZ  Message
0        started  10.219.211.94  juju-509789-0  ubuntu@24.04      Running

$ juju add-unit ubuntu
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m      local       localhost/localhost  4.0-beta5.1  21:04:50Z

App     Version  Status  Scale  Charm   Channel  Rev  Exposed  Message
ubuntu           active      2  ubuntu             0  no       

Unit       Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*  active    idle   0        10.219.211.94          
ubuntu/1   active    idle   1        10.219.211.198         

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.94   juju-509789-0  ubuntu@24.04      Running
1        started  10.219.211.198  juju-509789-1  ubuntu@24.04      Running
```

### Migrate

```
$ juju bootstrap lxd dst
$ juju bootstrap lxd src
$ juju add-model m
$ juju deploy ubuntu ch
$ juju deploy ./ubuntu_r25.charm
(wait until ready)
$ juju migrate m dst
(verify migration successful)
```

## Links

**Jira card:** https://warthogs.atlassian.net/browse/JUJU-6842

### CODE REVIEW

File: apiserver/charms.go [hunk 1]
```diff
@@ -1,51 +0,0 @@
...
-		return archive, bytes.Buffer{}, "", errors.Errorf("creating temp directory: %w", err)
-	}
-	defer os.RemoveAll(tempDir)
-	extractPath := filepath.Join(tempDir, "extracted")
-
-	// Expand and repack it with the specified revision
-	archive.SetRevision(charmRevision)
-	if err := archive.ExpandTo(extractPath); err != nil {
-		return archive, bytes.Buffer{}, "", errors.Errorf("extracting uploaded charm: %w", err)
-	}
```
Comment (SimonRichardson) [inline apiserver/charms.go:34]: Once this was skipped down, it made it obvious what could be removed.

**Review by nvinuesa — APPROVED** [review]
Nice cleanup! QA went well.

File: apiserver/charms.go (hunk 1 above)
Comment (jack-w-shaw) [reply apiserver/charms.go:34]: I'm _very_ happy to see this go!

File: internal/charm/charmdir.go [hunk 2]
```diff
@@ -170,7 +170,7 @@ func (dir *CharmDir) join(parts ...string) string {
 // SetDiskRevision does the same as SetRevision but also changes
 // the revision file in the charm directory.
 func (dir *CharmDir) SetDiskRevision(revision int) error {
-	dir.SetRevision(revision)
+	dir.setRevision(revision)
```
Comment (jack-w-shaw) [inline internal/charm/charmdir.go:173]: This method is only actually used in tests. If we want charms to be truly immutable, it should be moved off of `ChamrDir` so we aren't tempted to use it future. Perhaps a helper function in a `testing` package? What do you think?

**Review by jack-w-shaw — COMMENTED** [review]
One small request

File: internal/charm/charmdir.go (hunk 2 above)
Comment (SimonRichardson) [reply internal/charm/charmdir.go:173]: Correct, I don't think it's worth it right now, but we can evaluate this package to clean it up.

**Comment by jack-w-shaw** [comment]
It could do with some more QA steps
```
$ juju add-model m
$ juju download ubuntu
$ juju deploy ./ubuntu_r25.charm app1
Located local charm "ubuntu", revision 0
Deploying "app1" from local charm "ubuntu", revision 0 on ubuntu@24.04/stable

$ juju deploy ./ubuntu_r25.charm app2
Located local charm "ubuntu", revision 1
Deploying "app2" from local charm "ubuntu", revision 1 on ubuntu@24.04/stable

$ juju status
$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m      lxd         localhost/localhost  4.0-beta5.1  11:48:23Z

App   Version  Status  Scale  Charm   Channel  Rev  Exposed  Message
app1  24.04    active      1  ubuntu             0  no       
app2  24.04    active      1  ubuntu             1  no       

Unit     Workload  Agent  Machine  Public address  Ports  Message
app1/0*  active    idle   0        10.219.211.86          
app2/0*  active    idle   1        10.219.211.234         

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.86   juju-440822-0  ubuntu@24.04      Running
1        started  10.219.211.234  juju-440822-1  ubuntu@24.04      Running
``` 
^ note that the revision increases

```
$ juju add-model m2
$ mkdir ch
$ cd ch
$ juju download ubuntu
$ unzip *
$ rm ubuntu_r25.charm
$ echo "10" > revision
$ zip -r ubuntu_r10.charm
$ juju deploy ./ubuntu_r10.charm
Located local charm "ubuntu", revision 10
Deploying "ubuntu" from local charm "ubuntu", revision 10 on ubuntu@24.04/stable

$ juju status
Model  Controller  Cloud/Region         Version      Timestamp
m2     lxd         localhost/localhost  4.0-beta5.1  12:16:08Z

App      Version  Status  Scale  Charm   Channel  Rev  Exposed  Message
ubuntu   24.04    active      1  ubuntu            10  no       

Unit        Workload  Agent  Machine  Public address  Ports  Message
ubuntu/0*   active    idle   0        10.219.211.203         

Machine  State    Address         Inst id        Base          AZ  Message
0        started  10.219.211.203  juju-59edf7-0  ubuntu@24.04      Running
```
^ note `ubuntu` deploys to revision 10

File: internal/charm/charmdir.go (hunk 2 above)
Comment (jack-w-shaw) [reply internal/charm/charmdir.go:173]: I agree

**Review by jack-w-shaw — APPROVED** [review]

**Pull request merged**